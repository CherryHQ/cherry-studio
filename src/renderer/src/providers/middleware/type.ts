import { MCPTool, MCPToolResponse, Metrics, Usage, WebSearchResponse } from '@renderer/types'
import { Chunk, ErrorChunk } from '@renderer/types/chunk'

import { BaseApiClient, SdkParams, SdkRawChunk, SdkRawOutput, SdkToolCall } from '../AiProvider/clients'
import { CompletionsParams, GenericChunk, OnFilterMessagesFunction } from './schemas'

/**
 * Symbol to uniquely identify middleware context objects.
 */
export const MIDDLEWARE_CONTEXT_SYMBOL = Symbol.for('AiProviderMiddlewareContext')

/**
 * Defines the structure for the onChunk callback function.
 */
export type OnChunkFunction = (chunk: Chunk | ErrorChunk) => void

/**
 * Base context that carries information about the current method call.
 */
export interface BaseContext {
  [MIDDLEWARE_CONTEXT_SYMBOL]: true
  methodName: string
  apiClientInstance: BaseApiClient
  originalParams: Readonly<CompletionsParams>

  readonly onChunkCallback: (chunk: Chunk) => void
  readonly onFilterMessagesCallback?: OnFilterMessagesFunction // 可选
}

/**
 * Processing state shared between middlewares.
 */
export interface ProcessingState {
  sdkPayload?: Readonly<SdkParams>
  capabilities?: {
    isStreaming: boolean
    isEnabledToolCalling: boolean
    isEnabledWebSearch: boolean
    isEnabledReasoning: boolean
    mcpTools: MCPTool[]
  }
  observer?: {
    usage?: Usage
    metrics?: Metrics
  }
  apiCall?: {
    requestTimestamp?: number
    rawSdkOutput?: SdkRawOutput
    rawSdkStream?: ReadableStream<SdkRawChunk> // Output from StreamAdapterMiddleware, consumed by SdkChunkToGenericChunkMiddleware
    genericChunkStream?: ReadableStream<GenericChunk> // Output from SdkChunkToGenericChunkMiddleware, consumed by Generic processors (Text, Think, Image, etc.)
  }
  toolProcessingState?: {
    pendingToolCalls: Array<SdkToolCall>
    executingToolCalls: Array<{
      sdkToolCall: SdkToolCall
      mcpToolResponse: MCPToolResponse
    }>
    isRecursiveCall?: boolean
    recursionDepth?: number
  }
  webSearchState?: {
    results?: WebSearchResponse[]
  }
  flowControl?: {
    abortController?: AbortController
    abortSignal?: AbortSignal
    cleanup?: () => void
  }
  customState?: Record<string, any>
}

/**
 * Extended context for completions method.
 */
export interface CompletionsContext extends BaseContext {
  readonly methodName: 'completions' // 强制方法名为 'completions'

  // --- Mutable internal state for the duration of the middleware chain ---
  _internal: ProcessingState // 包含所有可变的处理状态
}

/**
 * Next function type in Koa style.
 */
export type Next = () => Promise<void>

/**
 * Base middleware type.
 */
export type Middleware<TContext extends BaseContext> = (ctx: TContext, next: Next) => Promise<void>

/**
 * Completions middleware type.
 */
export type CompletionsMiddleware = Middleware<CompletionsContext>

/**
 * Generic method middleware type.
 */
export type MethodMiddleware = Middleware<BaseContext>

/**
 * Base configuration for any middleware.
 */
export interface BaseMiddlewareConfig {
  id?: string
  name?: string
}

/**
 * Middleware configuration structure.
 */
export interface MiddlewareConfig extends BaseMiddlewareConfig {
  completions?: CompletionsMiddleware[]
  methods?: Record<string, MethodMiddleware[]>
}

// Re-export for convenience
export type { Chunk as OnChunkArg } from '@renderer/types/chunk'
