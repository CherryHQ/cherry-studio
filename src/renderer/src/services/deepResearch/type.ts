import { z } from 'zod'

// Schema
// =====================================================
export const ClarifyWithUserSchema = z
  .object({
    need_clarification: z.boolean().describe('Whether the query needs clarification from the user'),
    question: z.string().describe('question to ask the user to clarify the report scope'),
    verification: z.string().describe('verification message that we will start research')
  })
  .transform((r) => ({
    needClarification: r.need_clarification,
    question: r.question,
    verification: r.verification
  }))
export type ClarifyWithUserResponse = z.infer<typeof ClarifyWithUserSchema>

export const ResearchTopicsSchema = z.object({
  reflection: z.string().describe('thorough reflection on the research brief and current findings'),
  tasks: z.array(z.string()).describe('list of research tasks to explore'),
  fulfilled: z.boolean().describe('whether the research brief is fulfilled by the current findings')
})
export type ResearchTopicsResponse = z.infer<typeof ResearchTopicsSchema>
// =====================================================
