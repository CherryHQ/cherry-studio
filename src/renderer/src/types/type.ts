export type Maybe<T> = T | undefined

/**
 * 获取对象的所有键名，并保持类型安全
 * @param obj - 要获取键名的对象
 * @returns 对象的所有键名数组，类型为对象键名的联合类型
 * @example
 * ```ts
 * const obj = { foo: 1, bar: 'hello' };
 * const keys = objectKeys(obj); // ['foo', 'bar']
 * ```
 */
export function objectKeys<T extends object>(obj: T): (keyof T)[] {
  return Object.keys(obj) as (keyof T)[]
}

/**
 * 将对象转换为键值对数组，保持类型安全
 * @template T - 对象类型
 * @param obj - 要转换的对象
 * @returns 键值对数组，每个元素是一个包含键和值的元组
 * @example
 * const obj = { name: 'John', age: 30 };
 * const entries = objectEntries(obj); // [['name', 'John'], ['age', 30]]
 */
export function objectEntries<T extends object>(obj: T): [keyof T, T[keyof T]][] {
  return Object.entries(obj) as [keyof T, T[keyof T]][]
}

/**
 * 将对象转换为键值对数组，提供更严格的类型检查
 * @template T - 对象类型，键必须是string、number或symbol，值可以是任意类型
 * @param obj - 要转换的对象
 * @returns 键值对数组，每个元素是一个包含键和值的元组，类型完全对应原对象的键值类型
 * @example
 * const obj = { name: 'John', age: 30 };
 * const entries = objectEntriesStrict(obj); // [['name', string], ['age', number]]
 */
export function objectEntriesStrict<T extends Record<string | number | symbol, unknown>>(
  obj: T
): { [K in keyof T]: [K, T[K]] }[keyof T][] {
  return Object.entries(obj) as { [K in keyof T]: [K, T[K]] }[keyof T][]
}

/**
 * 获取对象所有值的类型安全版本
 * @template T - 对象类型
 * @param obj - 要获取值的对象
 * @returns 对象值组成的数组
 * @example
 * const obj = { a: 1, b: 2 } as const;
 * const values = objectValues(obj); // (1 | 2)[]
 */
export function objectValues<T extends Record<string, unknown>>(obj: T): T[keyof T][] {
  return Object.values(obj) as T[keyof T][]
}

/**
 * 表示一个对象类型，该对象至少包含类型T中指定的所有键，这些键的值类型为U
 * 同时也允许包含其他任意string类型的键，这些键的值类型也必须是U
 * @template T - 必需包含的键的字面量字符串联合类型
 * @template U - 所有键对应值的类型
 * @example
 * type Example = AtLeast<'a' | 'b', number>;
 * // 结果类型允许:
 * const obj1: Example = { a: 1, b: 2 };           // 只包含必需的键
 * const obj2: Example = { a: 1, b: 2, c: 3 };     // 包含额外的键
 */
export type AtLeast<T extends string, U> = {
  [K in T]: U
} & {
  [key: string]: U
}

/**
 * 从对象中移除指定的属性键，返回新对象
 * @template T - 源对象类型
 * @template K - 要移除的属性键类型，必须是T的键
 * @param obj - 源对象
 * @param keys - 要移除的属性键列表
 * @returns 移除指定属性后的新对象
 * @example
 * ```ts
 * const obj = { a: 1, b: 2, c: 3 };
 * const result = strip(obj, ['a', 'b']);
 * // result = { c: 3 }
 * ```
 */
export function strip<T, K extends keyof T>(obj: T, keys: K[]): Omit<T, K> {
  const result = { ...obj }
  for (const key of keys) {
    delete (result as any)[key] // 类型上 Omit 已保证安全
  }
  return result
}

/**
 * Makes specified properties required while keeping others as is
 * @template T - The object type to modify
 * @template K - Keys of T that should be required
 * @example
 * type User = {
 *   name?: string;
 *   age?: number;
 * }
 *
 * type UserWithName = RequireSome<User, 'name'>
 * // Result: { name: string; age?: number; }
 */
export type RequireSome<T, K extends keyof T> = Omit<T, K> & Required<Pick<T, K>>
