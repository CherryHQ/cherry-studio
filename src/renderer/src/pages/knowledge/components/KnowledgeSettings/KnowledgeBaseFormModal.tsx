import { Button, Dialog, DialogContent, DialogFooter, DialogHeader, DialogTitle, RowFlex } from '@cherrystudio/ui'
import { Menu } from 'antd'
import React, { useState } from 'react'
import { useTranslation } from 'react-i18next'
import styled from 'styled-components'

export interface PanelConfig {
  key: string
  label: string
  panel: React.ReactNode
}

interface KnowledgeBaseFormModalProps {
  title?: string
  open: boolean
  onCancel: () => void
  onOk?: () => void
  panels: PanelConfig[]
}

const KnowledgeBaseFormModal: React.FC<KnowledgeBaseFormModalProps> = ({ title, open, onCancel, onOk, panels }) => {
  const { t } = useTranslation()
  const [selectedMenu, setSelectedMenu] = useState(panels[0]?.key)

  const menuItems = panels.map(({ key, label }) => ({ key, label }))
  const activePanel = panels.find((p) => p.key === selectedMenu)?.panel

  const handleOpenChange = (newOpen: boolean) => {
    if (!newOpen) {
      onCancel()
    }
  }

  return (
    <Dialog open={open} onOpenChange={handleOpenChange}>
      <DialogContent
        showCloseButton={true}
        onPointerDownOutside={(e) => e.preventDefault()}
        className="w-[900px] max-w-[65vw] overflow-hidden p-0 pb-[10px] sm:max-w-[900px]">
        {title && (
          <DialogHeader className="m-0 rounded-none border-[var(--color-border)] border-b-[0.5px] p-[10px_15px]">
            <DialogTitle className="text-sm">{title}</DialogTitle>
          </DialogHeader>
        )}
        <RowFlex className="h-[550px]">
          <LeftMenu>
            <StyledMenu
              defaultSelectedKeys={[selectedMenu]}
              mode="vertical"
              items={menuItems}
              onSelect={({ key }) => setSelectedMenu(key)}
            />
          </LeftMenu>
          <SettingsContentPanel>{activePanel}</SettingsContentPanel>
        </RowFlex>
        {onOk && (
          <DialogFooter className="border-[var(--color-border)] border-t-[0.5px] px-4 pt-3">
            <Button variant="bordered" onPress={onCancel}>
              {t('common.cancel')}
            </Button>
            <Button onPress={onOk}>{t('common.confirm')}</Button>
          </DialogFooter>
        )}
      </DialogContent>
    </Dialog>
  )
}

const LeftMenu = styled.div`
  display: flex;
  height: 100%;
  border-right: 0.5px solid var(--color-border);
`

const SettingsContentPanel = styled.div`
  flex: 1;
  padding: 16px 16px;
  overflow-y: scroll;
`

const StyledMenu = styled(Menu)`
  width: 200px;
  padding: 5px;
  background: transparent;
  margin-top: 2px;
  border-inline-end: none !important;

  .ant-menu-item {
    height: 36px;
    color: var(--color-text-2);
    display: flex;
    align-items: center;
    border: 0.5px solid transparent;
    border-radius: 6px;
    margin-bottom: 7px;

    .ant-menu-title-content {
      line-height: 36px;
    }
  }
  .ant-menu-item-active {
    background-color: var(--color-background-soft) !important;
    transition: none;
  }
  .ant-menu-item-selected {
    background-color: var(--color-background-soft);
    border: 0.5px solid var(--color-border);
    .ant-menu-title-content {
      color: var(--color-text-1);
      font-weight: 500;
    }
  }
`

export default KnowledgeBaseFormModal
