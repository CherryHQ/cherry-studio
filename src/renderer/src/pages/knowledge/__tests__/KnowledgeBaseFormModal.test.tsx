import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import React from "react";
import { beforeEach, describe, expect, it, vi } from "vitest";

import type { PanelConfig } from "../components/KnowledgeSettings/KnowledgeBaseFormModal";
import KnowledgeBaseFormModal from "../components/KnowledgeSettings/KnowledgeBaseFormModal";

// Mock dependencies
const mocks = vi.hoisted(() => ({
  onCancel: vi.fn(),
  onOk: vi.fn(),
  t: vi.fn((key: string) => {
    const translations: Record<string, string> = {
      "common.cancel": "Cancel",
      "common.confirm": "Confirm",
    };
    return translations[key] || key;
  }),
}));

// Mock @cherrystudio/ui components (shadcn)
vi.mock("@cherrystudio/ui", () => ({
  Dialog: ({ children, open, onOpenChange }: any) =>
    open ? (
      <div data-testid="dialog" data-open={open}>
        {children}
      </div>
    ) : null,
  DialogContent: ({
    children,
    showCloseButton,
    onPointerDownOutside,
    className,
  }: any) => (
    <div
      data-testid="dialog-content"
      className={className}
      data-show-close={showCloseButton}
      onPointerDown={onPointerDownOutside}
    >
      {children}
    </div>
  ),
  DialogHeader: ({ children, className }: any) => (
    <div data-testid="dialog-header" className={className}>
      {children}
    </div>
  ),
  DialogTitle: ({ children, className }: any) => (
    <h2 data-testid="dialog-title" className={className}>
      {children}
    </h2>
  ),
  DialogFooter: ({ children, className }: any) => (
    <div data-testid="dialog-footer" className={className}>
      {children}
    </div>
  ),
  Button: ({ children, onPress, variant, ...props }: any) => (
    <button
      type="button"
      data-testid="button"
      onClick={onPress}
      data-variant={variant}
      {...props}
    >
      {children}
    </button>
  ),
  Separator: ({ orientation }: any) => (
    <div data-testid="separator" data-orientation={orientation} />
  ),
  Tabs: ({ children, defaultValue, orientation, className }: any) => (
    <div
      data-testid="tabs"
      data-default-value={defaultValue}
      data-orientation={orientation}
      className={className}
    >
      {children}
    </div>
  ),
  TabsList: ({ children, className }: any) => (
    <div data-testid="tabs-list" className={className}>
      {children}
    </div>
  ),
  TabsTrigger: ({ children, value, className, ...props }: any) => (
    <button
      type="button"
      data-testid={`tabs-trigger-${value}`}
      data-value={value}
      className={className}
      {...props}
    >
      {children}
    </button>
  ),
  TabsContent: ({ children, value, className }: any) => (
    <div
      data-testid={`tabs-content-${value}`}
      data-value={value}
      className={className}
    >
      {children}
    </div>
  ),
}));

// Mock react-i18next
vi.mock("react-i18next", () => ({
  useTranslation: () => ({ t: mocks.t }),
}));

/**
 * 创建测试用的面板配置
 * @param overrides 可选的属性覆盖
 * @returns PanelConfig 数组
 */
function createPanelConfigs(
  overrides: Partial<PanelConfig>[] = [],
): PanelConfig[] {
  const defaultPanels: PanelConfig[] = [
    {
      key: "general",
      label: "General Settings",
      panel: <div data-testid="general-panel">General Settings Panel</div>,
    },
    {
      key: "advanced",
      label: "Advanced Settings",
      panel: <div data-testid="advanced-panel">Advanced Settings Panel</div>,
    },
  ];

  return defaultPanels.map((panel, index) => ({
    ...panel,
    ...overrides[index],
  }));
}

/**
 * 渲染 KnowledgeBaseFormModal 组件的辅助函数
 * @param props 可选的组件属性
 * @returns render 结果
 */
function renderModal(props: Partial<any> = {}) {
  const defaultProps = {
    open: true,
    title: "Knowledge Base Settings",
    panels: createPanelConfigs(),
    onCancel: mocks.onCancel,
    onOk: mocks.onOk,
  };

  return render(<KnowledgeBaseFormModal {...defaultProps} {...props} />);
}

describe("KnowledgeBaseFormModal", () => {
  beforeEach(() => {
    vi.clearAllMocks();
  });

  describe("basic rendering", () => {
    it("should match snapshot", () => {
      const { container } = renderModal();
      expect(container.firstChild).toMatchSnapshot();
    });

    it("should render modal when open is true", () => {
      renderModal({ open: true });

      expect(screen.getByTestId("dialog")).toBeInTheDocument();
      expect(screen.getByTestId("dialog-content")).toBeInTheDocument();
      expect(screen.getByTestId("tabs")).toBeInTheDocument();
    });

    it("should render first panel by default", () => {
      renderModal();

      // shadcn Tabs renders all TabsContent (unlike antd Menu)
      // Both panels should be in the document
      expect(screen.getByTestId("general-panel")).toBeInTheDocument();
      expect(screen.getByTestId("advanced-panel")).toBeInTheDocument();
    });

    it("should handle empty panels array", () => {
      renderModal({ panels: [] });

      expect(screen.getByTestId("dialog")).toBeInTheDocument();
      expect(screen.getByTestId("tabs")).toBeInTheDocument();
    });
  });

  describe("tabs interaction", () => {
    it("should render all tabs content", () => {
      renderModal();

      // All tab contents should be rendered (shadcn tabs renders all content)
      expect(screen.getByTestId("tabs-content-general")).toBeInTheDocument();
      expect(screen.getByTestId("tabs-content-advanced")).toBeInTheDocument();
    });

    it("should set default value to first panel key", () => {
      const panels = createPanelConfigs();
      renderModal({ panels });

      const tabs = screen.getByTestId("tabs");
      expect(tabs).toHaveAttribute("data-default-value", panels[0].key);
    });

    it("should render tabs with custom panels", () => {
      const customPanels: PanelConfig[] = [
        {
          key: "custom1",
          label: "Custom Panel 1",
          panel: <div data-testid="custom1-panel">Custom Panel 1</div>,
        },
        {
          key: "custom2",
          label: "Custom Panel 2",
          panel: <div data-testid="custom2-panel">Custom Panel 2</div>,
        },
      ];

      renderModal({ panels: customPanels });

      // Both custom panels should be rendered
      expect(screen.getByTestId("custom1-panel")).toBeInTheDocument();
      expect(screen.getByTestId("custom2-panel")).toBeInTheDocument();

      // Tabs triggers should exist
      expect(screen.getByTestId("tabs-trigger-custom1")).toBeInTheDocument();
      expect(screen.getByTestId("tabs-trigger-custom2")).toBeInTheDocument();
    });
  });

  describe("dialog props", () => {
    const user = userEvent.setup();
    it("should display title correctly", () => {
      const customTitle = "Custom Modal Title";
      renderModal({ title: customTitle });

      const title = screen.getByTestId("dialog-title");
      expect(title).toHaveTextContent(customTitle);
    });

    it("should call onOk when confirm button is clicked", async () => {
      renderModal();

      const buttons = screen.getAllByTestId("button");
      const confirmButton = buttons.find(
        (btn) => btn.textContent === "Confirm",
      );

      if (confirmButton) {
        await user.click(confirmButton);
        expect(mocks.onOk).toHaveBeenCalledTimes(1);
      }
    });
  });

  describe("edge cases", () => {
    it("should handle single panel", () => {
      const singlePanel: PanelConfig[] = [
        {
          key: "only",
          label: "Only Panel",
          panel: <div data-testid="only-panel">Only Panel</div>,
        },
      ];

      renderModal({ panels: singlePanel });

      expect(screen.getByTestId("only-panel")).toBeInTheDocument();
      expect(screen.getByTestId("tabs-trigger-only")).toBeInTheDocument();
    });

    it("should handle panel with undefined key gracefully", () => {
      const panelsWithUndefined = [
        {
          key: "valid",
          label: "Valid Panel",
          panel: <div data-testid="valid-panel">Valid Panel</div>,
        },
      ];

      renderModal({ panels: panelsWithUndefined });

      expect(screen.getByTestId("valid-panel")).toBeInTheDocument();
    });
  });
});
