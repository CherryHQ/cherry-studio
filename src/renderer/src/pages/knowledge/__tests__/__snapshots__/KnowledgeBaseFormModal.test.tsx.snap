// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`KnowledgeBaseFormModal > basic rendering > should match snapshot 1`] = `
<div
  data-open="true"
  data-testid="dialog"
>
  <div
    class="w-[900px] max-w-[65vw] overflow-hidden p-0 pb-[10px] sm:max-w-[900px]"
    data-show-close="true"
    data-testid="dialog-content"
  >
    <div
      class="m-0 rounded-none border-[var(--color-border)] border-b-[0.5px] p-[10px_15px]"
      data-testid="dialog-header"
    >
      <h2
        class="text-sm"
        data-testid="dialog-title"
      >
        Knowledge Base Settings
      </h2>
    </div>
    <div
      class="flex h-[550px] flex-row"
      data-default-value="general"
      data-orientation="vertical"
      data-testid="tabs"
    >
      <div
        class="flex h-full w-auto items-start justify-center bg-transparent px-2"
        data-testid="tabs-list"
      >
        <div
          class="mt-[2px] h-auto flex-col gap-0 bg-transparent p-[5px]"
          data-testid="tabs-list"
        >
          <button
            class="mb-[7px] flex h-9 items-center justify-start rounded-md border-[0.5px] border-transparent bg-transparent px-3 font-normal text-[var(--color-text-2)] shadow-none hover:bg-[var(--color-background-soft)] data-[state=active]:border-[var(--color-border)] data-[state=active]:bg-[var(--color-background-soft)] data-[state=active]:font-medium data-[state=active]:text-[var(--color-text-1)] data-[state=active]:shadow-none"
            data-testid="tabs-trigger-general"
            data-value="general"
            type="button"
          >
            General Settings
          </button>
          <button
            class="mb-[7px] flex h-9 items-center justify-start rounded-md border-[0.5px] border-transparent bg-transparent px-3 font-normal text-[var(--color-text-2)] shadow-none hover:bg-[var(--color-background-soft)] data-[state=active]:border-[var(--color-border)] data-[state=active]:bg-[var(--color-background-soft)] data-[state=active]:font-medium data-[state=active]:text-[var(--color-text-1)] data-[state=active]:shadow-none"
            data-testid="tabs-trigger-advanced"
            data-value="advanced"
            type="button"
          >
            Advanced Settings
          </button>
        </div>
      </div>
      <div
        data-orientation="vertical"
        data-testid="separator"
      />
      <div
        class="flex-1 overflow-y-auto"
        data-testid="tabs-content-general"
        data-value="general"
      >
        <div
          data-testid="general-panel"
        >
          General Settings Panel
        </div>
      </div>
      <div
        class="flex-1 overflow-y-auto"
        data-testid="tabs-content-advanced"
        data-value="advanced"
      >
        <div
          data-testid="advanced-panel"
        >
          Advanced Settings Panel
        </div>
      </div>
    </div>
    <div
      class="border-[var(--color-border)] border-t-[0.5px] px-4 pt-3"
      data-testid="dialog-footer"
    >
      <button
        data-testid="button"
        data-variant="bordered"
        type="button"
      >
        Cancel
      </button>
      <button
        data-testid="button"
        type="button"
      >
        Confirm
      </button>
    </div>
  </div>
</div>
`;
